From d2ae7029fbad21dfd03ad04dd0ba8a8457d04dcd Mon Sep 17 00:00:00 2001
From: Gabriel Goller <g.goller@proxmox.com>
Date: Tue, 25 Aug 2026 14:41:21 +0200
Subject: [PATCH] zebra: release route node after connected cleanup

route_node_match() returns a locked route node, but the node is never
released. Jump to the release label on every return.

Signed-off-by: Gabriel Goller <g.goller@proxmox.com>
---
 zebra/connected.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/zebra/connected.c b/zebra/connected.c
index a288a327ac32..4eb476e59783 100644
--- a/zebra/connected.c
+++ b/zebra/connected.c
@@ -205,18 +205,21 @@ static void connected_remove_kernel_for_connected(afi_t afi, safi_t safi, struct
 		return;
 
 	if (!prefix_same(&rn->p, p))
-		return;
+		goto done;
 
 	dest = rib_dest_from_rnode(rn);
 	if (!dest || !dest->selected_fib)
-		return;
+		goto done;
 
 	re = dest->selected_fib;
 	if (re->type != ZEBRA_ROUTE_KERNEL)
-		return;
+		goto done;
 
 	rib_delete(afi, SAFI_UNICAST, zvrf->vrf->vrf_id, ZEBRA_ROUTE_KERNEL, re->instance,
 		   re->flags, p, NULL, nh, 0, zvrf->table_id, re->metric, re->distance, false);
+
+done:
+	route_unlock_node(rn);
 }
 
 /* Called from if_up(). */
